src/base/bittorrent/dbresumedatastorage.cpp

```diff
@@ -0,0 +1,573 @@
+            makeColumnDefinition(DB_COLUMN_HAS_SEED_STATUS, "INTEGER NOT NULL"),
+            makeColumnDefinition(DB_COLUMN_OPERATING_MODE, "TEXT NOT NULL"),
+            makeColumnDefinition(DB_COLUMN_STOPPED, "INTEGER NOT NULL"),
+            makeColumnDefinition(DB_COLUMN_RESUMEDATA, "BLOB NOT NULL"),
+            makeColumnDefinition(DB_COLUMN_METADATA, "BLOB")
+        };
+        const QString createTableTorrentsQuery = makeCreateTableStatement(DB_TABLE_TORRENTS, tableTorrentsItems);
+        if (!query.exec(createTableTorrentsQuery))
+            throw RuntimeError(query.lastError().text());
+
+        if (!db.commit())
+            throw RuntimeError(db.lastError().text());
+    }
+    catch (const RuntimeError &err)
+    {
+        db.rollback();
+        throw err;
+    }
+}
+
+BitTorrent::DBResumeDataStorage::Worker::Worker(const QString &dbPath, const QString &dbConnectionName)
+    : m_path {dbPath}
+    , m_connectionName {dbConnectionName}
+{
+}
+
+void BitTorrent::DBResumeDataStorage::Worker::openDatabase() const
+{
+    auto db = QSqlDatabase::addDatabase(QLatin1String("QSQLITE"), m_connectionName);
+    db.setDatabaseName(m_path);
+    if (!db.open())
+        throw RuntimeError(db.lastError().text());
+}
+
+void BitTorrent::DBResumeDataStorage::Worker::closeDatabase() const
+{
+    QSqlDatabase::removeDatabase(m_connectionName);
+}
+
+void BitTorrent::DBResumeDataStorage::Worker::store(const TorrentID &id, const LoadTorrentParams &resumeData) const
+{
+    // We need to adjust native libtorrent resume data
+    lt::add_torrent_params p = resumeData.ltAddTorrentParams;
+    p.save_path = Profile::instance()->toPortablePath(QString::fromStdString(p.save_path)).toStdString();
+    if (resumeData.stopped)
+    {
+        p.flags |= lt::torrent_flags::paused;
+        p.flags &= ~lt::torrent_flags::auto_managed;
+    }
+    else
+    {
+        // Torrent can be actually "running" but temporarily "paused" to perform some
+        // service jobs behind the scenes so we need to restore it as "running"
+        if (resumeData.operatingMode == BitTorrent::TorrentOperatingMode::AutoManaged)
+        {
+            p.flags |= lt::torrent_flags::auto_managed;
+        }
+        else
+        {
+            p.flags &= ~lt::torrent_flags::paused;
+            p.flags &= ~lt::torrent_flags::auto_managed;
+        }
+    }
+
+    auto columns = QVector<Column>({
+        DB_COLUMN_TORRENT_ID,
+        DB_COLUMN_NAME,
+        DB_COLUMN_CATEGORY,
+        DB_COLUMN_TAGS,
+        DB_COLUMN_TARGET_SAVE_PATH,
+        DB_COLUMN_CONTENT_LAYOUT,
+        DB_COLUMN_RATIO_LIMIT,
+        DB_COLUMN_SEEDING_TIME_LIMIT,
+        DB_COLUMN_HAS_OUTER_PIECES_PRIORITY,
+        DB_COLUMN_HAS_SEED_STATUS,
+        DB_COLUMN_OPERATING_MODE,
+        DB_COLUMN_STOPPED,
+        DB_COLUMN_RESUMEDATA
+    });
+
+    // metadata is stored in separate column
+    QByteArray bencodedMetadata;
```

Comment:
IMO it is still worth to reserve some memory, even some approximate value is better.
Same for `bencodedResumeData`.

Reply:
>IMO it is still worth to reserve some memory, even some approximate value is better.

I wonder if there is some reasonable way to get an approximate size of resulting data based on source `lt::torrent_info`?

Reply:
Maybe make buffers static so we'll avoid reallocations?

Reply:
>Maybe make buffers static so we'll avoid reallocations?

This is a bit too much IMO.

>I wonder if there is some reasonable way to get an approximate size of resulting data based on source lt::torrent_info?

IIRC I asked for something similar but the answer was it is impossible to give an estimate without traversing the whole data structure.
Maybe you could measure the average size of your testing workload and round it up to the nearest 2^n number. This way we at least save some relocations.

Reply:
>Maybe you could measure the average size of your testing workload and round it up to the nearest 2^n number. This way we at least save some relocations.

My use case is limited. I do not deal with large torrents, such as HD videos, etc.

The size of the torrent file and resume data largely depends on the number of pieces in the torrent, the number of trackers, the number of files. There doesn't seem to be a way to choose some common size. So we can only choose some reasonable value to simply reduce the amount of memory reallocation. I suggest 256kB for resume data and 512kB for torrent files. What do you say to that?

Reply:
>I suggest 256kB for resume data and 512kB for torrent files. What do you say to that?

OK, we can start from there and tweak it later if there is such need.